julia/src/BridgeStan.jl

```diff
@@ -0,0 +1,336 @@
+    param_unc_names,
+    param_constrain!,
+    param_unconstrain!,
+    param_unconstrain_json!,
+    log_density_gradient!,
+    log_density_hessian!,
+    param_constrain,
+    param_unconstrain,
+    param_unconstrain_json,
+    log_density,
+    log_density_gradient,
+    log_density_hessian
+
+mutable struct StanModelStruct end
+
+mutable struct StanModel
+    lib::Ptr{Nothing}
+    stanmodel::Ptr{StanModelStruct}
+    const data::String
+    const seed::UInt32
+    const chain_id::UInt32
+
+    function StanModel(stanlib_::String, datafile_::String = "", seed_ = 204, chain_id_ = 0)
+        seed = convert(UInt32, seed_)
+        chain_id = convert(UInt32, chain_id_)
+
+        if !isfile(stanlib_)
+            throw(SystemError("Dynamic library file not found"))
+        end
+
+        if datafile_ != "" && !isfile(datafile_)
+            throw(SystemError("Data file not found"))
+        end
+
+        lib = Libc.Libdl.dlopen(stanlib_)
+
+        stanmodel = ccall(
+            Libc.Libdl.dlsym(lib, "construct"),
+            Ptr{StanModelStruct},
+            (Cstring, UInt32, UInt32),
+            datafile_,
+            seed,
+            chain_id,
+        )
+        if stanmodel == C_NULL
+            error("could not construct model RNG")
+        end
+
+        sm = new(lib, stanmodel, datafile_, seed, chain_id)
+
+        function f(sm)
+            ccall(
+                Libc.Libdl.dlsym(sm.lib, "destruct"),
+                UInt32,
+                (Ptr{StanModelStruct},),
+                sm.stanmodel,
+            )
+        end
+
+        finalizer(f, sm)
+    end
+end
+
+function name(sm::StanModel)
+    cstr = ccall(
+        Libc.Libdl.dlsym(sm.lib, "name"),
+        Cstring,
+        (Ptr{StanModelStruct},),
+        sm.stanmodel,
+    )
+    unsafe_string(cstr)
+end
+
+function param_num(sm::StanModel; include_tp = false, include_gq = false)
+    ccall(
+        Libc.Libdl.dlsym(sm.lib, "param_num"),
+        Cint,
+        (Ptr{StanModelStruct}, Cint, Cint),
+        sm.stanmodel,
+        include_tp,
+        include_gq,
+    )
+end
+
+function param_unc_num(sm::StanModel)
+    ccall(
+        Libc.Libdl.dlsym(sm.lib, "param_unc_num"),
+        Cint,
+        (Ptr{StanModelStruct},),
+        sm.stanmodel,
+    )
+end
+
+function param_names(sm::StanModel; include_tp = false, include_gq = false)
+    cstr = ccall(
+        Libc.Libdl.dlsym(sm.lib, "param_names"),
+        Cstring,
+        (Ptr{StanModelStruct}, Cint, Cint),
+        sm.stanmodel,
+        include_tp,
+        include_gq,
+    )
+    [string(s) for s in split(unsafe_string(cstr), ',')]
+end
+
+function param_unc_names(sm::StanModel)
+    cstr = ccall(
+        Libc.Libdl.dlsym(sm.lib, "param_unc_names"),
+        Cstring,
+        (Ptr{StanModelStruct},),
+        sm.stanmodel,
+    )
+    [string(s) for s in split(unsafe_string(cstr), ',')]
+end
+
+function param_constrain!(
+    sm::StanModel,
+    theta_unc,
+    out::Vector{Float64};
```

Comment:
I honestly don't know what to type `theta_unc` and `out` as, or if typing them at all is best.

The issue is that both `theta_unc` and `out` can be any type `T` such that `T` is a subtype of `AbstractVector`, with elements of `Float64`, *and* the elements are contiguous in memory.  It's the contiguous in memory part that I don't know how to type.  The rest of it would be

```
function param_constrain!(sm::StanModel, theta_unc::T, out::T; ...) where {T <: AbstractVector{Float64}}
...
end
```

Thoughts?

Reply:
I'm not that familiar with Julia's more exotic types, can you give an example of something which is `<: AbstractVector{Float64}`, has contiguous memory, but is _not_ `Vector{Float64}`? I'd like to include a test with that thing if we expand this type.

This is also the kind of thing which we could change in a backwards-compatible way later on

Reply:
Yep, you're right.  Being stricter now, by forcing `Vector{Float64}`, will allow us to accept a broader set of types in the future in a backwards-compatible way, if we need/want to.  Good point.


For now we still want to explicitly avoid something like 

```
M = rand(3, 3);
@views M[1, :] # not contiguous
```
being passed in.

Will you add the type `Vector{Float64}` to the arguments `theta_unc`, `theta`, and `q`? I think all `out` arguments have it already.

Views, and less important for our use case ranges, are both `<: AbstractVector{Float64}`, have contiguous memory, and are *not* `Vector{Float64}`.  Try

```
x = rand(10);
typeof(@views x[1:2]) <: AbstractVector{Float64}
typeof(@views x[1:2])
y = 1.0:10.0;
typeof(y) <: AbstractVector{Float64}
typeof(y)
```

And for future reference, here's a [discussion of this issue](https://github.com/JuliaLang/julia/issues/10889) with a [temporary solution](https://github.com/JuliaIO/Blosc.jl/blob/25d663c607542cbebaea45542619726fec71bb5e/src/Blosc.jl#L58).

Reply:
As was just revealed to me by a test that broke when fixing those types, it seems that Julia also does the right thing when given a `Matrix{Float64}` to a C call. So really, down the road, we don't even need AbstractVector, but just AbstractArray and continuous?